votingpool/withdrawal.go

```diff
@@ -0,0 +1,939 @@
+// script) for a given transaction input. They should match the order of pubkeys
+// in the script and an empty RawSig should be used when the private key for a
+// pubkey is not known.
+type TxSigs [][]RawSig
+
+// RawSig represents one of the signatures included in the unlocking script of
+// inputs spending from P2SH UTXOs.
+type RawSig []byte
+
+// byAmount defines the methods needed to satisify sort.Interface to
+// sort a slice of OutputRequests by their amount.
+type byAmount []OutputRequest
+
+func (u byAmount) Len() int           { return len(u) }
+func (u byAmount) Less(i, j int) bool { return u[i].Amount < u[j].Amount }
+func (u byAmount) Swap(i, j int)      { u[i], u[j] = u[j], u[i] }
+
+// byOutBailmentID defines the methods needed to satisify sort.Interface to sort
+// a slice of OutputRequests by their outBailmentIDHash.
+type byOutBailmentID []OutputRequest
+
+func (s byOutBailmentID) Len() int      { return len(s) }
+func (s byOutBailmentID) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s byOutBailmentID) Less(i, j int) bool {
+	return bytes.Compare(s[i].outBailmentIDHash(), s[j].outBailmentIDHash()) < 0
+}
+
+func (s outputStatus) String() string {
+	strings := map[outputStatus]string{
+		statusSuccess: "success",
+		statusPartial: "partial-",
+		statusSplit:   "split",
+	}
+	return strings[s]
+}
+
+// Outputs returns a map of outbailment IDs to WithdrawalOutputs for all outputs
+// requested in this withdrawal.
+func (s *WithdrawalStatus) Outputs() map[OutBailmentID]*WithdrawalOutput {
+	return s.outputs
+}
+
+// Sigs returns a map of ntxids to signature lists for every input in the tx
+// with that ntxid.
+func (s *WithdrawalStatus) Sigs() map[Ntxid]TxSigs {
+	return s.sigs
+}
+
+// Fees returns the total amount of network fees included in all transactions
+// generated as part of a withdrawal.
+func (s *WithdrawalStatus) Fees() btcutil.Amount {
+	return s.fees
+}
+
+// NextInputAddr returns the votingpool address that should be used as the
+// startAddress of subsequent withdrawals.
+func (s *WithdrawalStatus) NextInputAddr() WithdrawalAddress {
+	return s.nextInputAddr
+}
+
+// NextChangeAddr returns the votingpool address that should be used as the
+// changeStart of subsequent withdrawals.
+func (s *WithdrawalStatus) NextChangeAddr() ChangeAddress {
+	return s.nextChangeAddr
+}
+
+// String makes OutputRequest satisfy the Stringer interface.
+func (r OutputRequest) String() string {
+	return fmt.Sprintf("OutputRequest %s to send %v to %s", r.outBailmentID(), r.Amount, r.Address)
+}
+
+func (r OutputRequest) outBailmentID() OutBailmentID {
+	return OutBailmentID(fmt.Sprintf("%s:%d", r.Server, r.Transaction))
+}
+
+// outBailmentIDHash returns a byte slice which is used when sorting
+// OutputRequests.
+func (r OutputRequest) outBailmentIDHash() []byte {
+	if r.cachedHash != nil {
+		return r.cachedHash
+	}
+	str := fmt.Sprintf("%s%d", r.Server, r.Transaction)
```

Comment:
Even though it's cached I made the change as it doesn't reduce readability IMO. Also good to know about fmt.Sprintf